### PR TITLE
Allow ArrayPool.Shared devirtualization

### PR DESCRIPTION
**Pull request by benaadams**
Every call to ArrayPool in crossgen'd System.Private.CoreLib is now devitualized to
```asm
call     TlsOverPerCoreLockedStacksArrayPool`1:Rent(int):ref:this
call     TlsOverPerCoreLockedStacksArrayPool`1:Return(ref,bool):this
```
Slight size regression
```asm
Total bytes of diff: 180 (0.01% of base)
    diff is a regression.

Total byte diff includes 178 bytes from reconciling methods
        Base had    0 unique methods,        0 unique bytes
        Diff had    3 unique methods,      178 unique bytes

Top file regressions by size (bytes):
         180 : System.Private.CoreLib.dasm (0.01% of base)

1 total files with size differences (0 improved, 1 regressed), 0 unchanged.

Top method regessions by size (bytes):
         110 : System.Private.CoreLib.dasm - ArrayPool:.cctor() (0/1 methods)
          34 : System.Private.CoreLib.dasm - ArrayPool:get_ByteArrayPool():ref (0/1 methods)
          34 : System.Private.CoreLib.dasm - ArrayPool:get_CharArrayPool():ref (0/1 methods)
          16 : System.Private.CoreLib.dasm - BinaryWriter:Write(struct):this (3 methods)
          13 : System.Private.CoreLib.dasm - ValueListBuilder`1:Grow():this (2 methods)

Top method improvements by size (bytes):
         -62 : System.Private.CoreLib.dasm - ArrayPool`1:.cctor() (4 methods)
         -21 : System.Private.CoreLib.dasm - Stream:ReadAsync(struct,struct):struct:this
         -13 : System.Private.CoreLib.dasm - Stream:Read(struct):int:this
         -13 : System.Private.CoreLib.dasm - TextReader:Read(struct):int:this
         -13 : System.Private.CoreLib.dasm - TextReader:ReadBlock(struct):int:this

36 total methods with size differences (5 improved, 31 regressed), 17261 unchanged.
```
Enabled by https://github.com/dotnet/coreclr/pull/15766

@stephentoub @jkotas PTAL

### CODE REVIEW

**Comment by benaadams** [comment]
/cc @AndyAyersMS 

```
NLINER: during 'fgInline' result 'success' reason 'aggressive inline attribute' for 'Stream:CopyTo(ref,int):this' calling 'ArrayPool`1:get_Shared():ref'
INLINER: during 'fgInline' result 'success' reason 'aggressive inline attribute'

Replacing the return expression placeholder [000011] with [000115]
               [000011] --C---------              *  RET_EXPR  ref   (inl return from call [000115])

Inserting the inline return expression
               [000115] ------------              *  LCL_VAR   ref    V07 tmp1         

**** Late devirt opportunity
               [000013] --C-G-------              *  CALLV ind ref    ArrayPool`1.Rent
               [000115] ------------ this in rcx  +--*  LCL_VAR   ref    V07 tmp1         
               [000012] ------------ arg1         \--*  LCL_VAR   int    V02 arg2         

impDevirtualizeCall: no type available (op=LCL_VAR)
Expanding INLINE_CANDIDATE in statement [000081] in BB08:
               [000081] ------------              *  STMT      void  (IL 0x040...  ???)
               [000080] I-C-G-------              \--*  CALL      ref    ArrayPool`1.get_Shared (exactContextHnd=0x000002408D3CBEB9)
INLINER: inlineInfo.tokenLookupContextHandle for ArrayPool`1:get_Shared():ref set to 0x000002408D3CBEB9:

Invoking compiler for the inlinee method ArrayPool`1:get_Shared():ref :
IL to import:
IL_0000  d0 a6 00 00 1b    ldtoken      0x1B0000A6
IL_0005  28 31 09 00 06    call         0x6000931
IL_000a  d0 a6 00 00 02    ldtoken      0x20000A6
IL_000f  28 31 09 00 06    call         0x6000931
IL_0014  28 38 09 00 06    call         0x6000938
IL_0019  2d 1b             brtrue.s     27 (IL_0036)
IL_001b  d0 a6 00 00 1b    ldtoken      0x1B0000A6
IL_0020  28 31 09 00 06    call         0x6000931
IL_0025  d0 a7 00 00 02    ldtoken      0x20000A7
IL_002a  28 31 09 00 06    call         0x6000931
IL_002f  28 38 09 00 06    call         0x6000938
IL_0034  2c 06             brfalse.s    6 (IL_003c)
IL_0036  7e 5c 04 00 0a    ldsfld       0xA00045C
IL_003b  2a                ret         
IL_003c  7e 5d 04 00 0a    ldsfld       0xA00045D
IL_0041  2a                ret         

INLINER impTokenLookupContextHandle for ArrayPool`1:get_Shared():ref is 0x000002408D3CBEB9.
*************** In fgFindBasicBlocks() for ArrayPool`1:get_Shared():ref
Jump targets:
  IL_0036
  IL_003c
New Basic Block BB19 [0016] created.
BB19 [000..01B)
New Basic Block BB20 [0017] created.
BB20 [01B..036)
New Basic Block BB21 [0018] created.
BB21 [036..03C)
New Basic Block BB22 [0019] created.
BB22 [03C..042)

lvaGrabTemp returning 8 (V08 tmp2) (a long lifetime temp) called for Inline return value spill temp.
Basic block list for 'ArrayPool`1:get_Shared():ref'

--------------------------------------------------------------------------------------------------------------------------------------
BBnum BBid ref try hnd                 weight    [IL range]     [jump]      [EH region]         [flags]
--------------------------------------------------------------------------------------------------------------------------------------
BB19 [0016]  1                             1     [000..01B)-> BB21 ( cond )                     
BB20 [0017]  1                             1     [01B..036)-> BB22 ( cond )                     
BB21 [0018]  2                             1     [036..03C)        (return)                     
BB22 [0019]  1                             1     [03C..042)        (return)                     
--------------------------------------------------------------------------------------------------------------------------------------
*************** In impImport() for ArrayPool`1:get_Shared():ref

impImportBlockPending for BB19

Importing BB19 (PC=000) of 'ArrayPool`1:get_Shared():ref'
    [ 0]   0 (0x000) ldtoken
    [ 1]   5 (0x005) call 06000931
In Compiler::impImportCall: opcode is call, kind=0, callRetType is ref, structSize is 0

    [ 1]  10 (0x00a) ldtoken
    [ 2]  15 (0x00f) call 06000931
In Compiler::impImportCall: opcode is call, kind=0, callRetType is ref, structSize is 0

    [ 2]  20 (0x014) call 06000938
In Compiler::impImportCall: opcode is call, kind=0, callRetType is bool, structSize is 0
Importing Type.op_*Equality intrinsic

Folding call to Type:op_Equality to a simple compare via EQ
Optimizing compare of types-from-handles to instead compare handles
Asking runtime to compare 000002408A69A650 (Byte) and 000002408A69A650 (Byte) for equality
Runtime reports comparison is known at jit time: 1

    [ 1]  25 (0x019) brtrue.s
Folding operator with constant nodes into a constant:
               [000126] ------------              /--*  CNS_INT   int    0
               [000127] ------------              *  NE        int   
               [000125] ------------              \--*  CNS_INT   int    1
Bashed to int constant:
               [000127] ------------              *  CNS_INT   int    1

The conditional jump becomes an unconditional jump to BB21

impImportBlockPending for BB21

Importing BB21 (PC=054) of 'ArrayPool`1:get_Shared():ref'
    [ 0]  54 (0x036) ldsfld 0A00045C
    [ 1]  59 (0x03b) ret

    Inlinee Return expression (before normalization)  =>
               [000137] --CXG-------              *  IND       ref   
               [000135] ------------              |  /--*  CNS_INT   int    0 Fseq[s_specificPool]
               [000136] --CXG-------              \--*  ADD       byref 
               [000134] H-CXG-------                 \--*  CALL help byref  HELPER.CORINFO_HELP_GETSHARED_GCSTATIC_BASE_DYNAMICCLASS
               [000130] #----------- arg0               +--*  IND       long  
               [000129] ------------                    |  \--*  CNS_INT(h) long   0x2408c12e7b0 cid/mid
               [000131] ------------ arg1               \--*  CNS_INT   int    58


               [000140] ------------              *  STMT      void  
               [000137] --CXG-------              |  /--*  IND       ref   
               [000135] ------------              |  |  |  /--*  CNS_INT   int    0 Fseq[s_specificPool]
               [000136] --CXG-------              |  |  \--*  ADD       byref 
               [000134] H-CXG-------              |  |     \--*  CALL help byref  HELPER.CORINFO_HELP_GETSHARED_GCSTATIC_BASE_DYNAMICCLASS
               [000130] #----------- arg0         |  |        +--*  IND       long  
               [000129] ------------              |  |        |  \--*  CNS_INT(h) long   0x2408c12e7b0 cid/mid
               [000131] ------------ arg1         |  |        \--*  CNS_INT   int    58
               [000139] -ACXG-------              \--*  ASG       ref   
               [000138] D------N----                 \--*  LCL_VAR   ref    V08 tmp2         


    Inlinee Return expression (after normalization) =>
               [000141] ------------              *  LCL_VAR   ref    V08 tmp2         

After impImport() added block for try,catch,finally
--------------------------------------------------------------------------------------------------------------------------------------
BBnum BBid ref try hnd                 weight    [IL range]     [jump]      [EH region]         [flags]
--------------------------------------------------------------------------------------------------------------------------------------
BB19 [0016]  1                             1     [000..01B)-> BB21 (always)                     i 
BB20 [0017]  1                             1     [01B..036)-> BB22 ( cond )                     
BB21 [0018]  2                             1     [036..03C)        (return)                     i 
BB22 [0019]  1                             1     [03C..042)        (return)                     
--------------------------------------------------------------------------------------------------------------------------------------

BB20 was not imported, marked as removed (1)
BB22 was not imported, marked as removed (2)

Renumbering the basic blocks for fgRemoveEmptyBlocks

*************** Before renumbering the basic blocks

--------------------------------------------------------------------------------------------------------------------------------------
BBnum BBid ref try hnd                 weight    [IL range]     [jump]      [EH region]         [flags]
--------------------------------------------------------------------------------------------------------------------------------------
BB19 [0016]  1                             1     [000..01B)-> BB21 (always)                     i 
BB21 [0018]  2                             1     [036..03C)        (return)                     i 
--------------------------------------------------------------------------------------------------------------------------------------

***************  Exception Handling table
index  eTry, eHnd
  0  ::            - Try at BB02..BB07 [016..038), Finally at BB08..BB08 [038..04D)
Renumber BB19 to BB23
Renumber BB21 to BB24

*************** After renumbering the basic blocks

--------------------------------------------------------------------------------------------------------------------------------------
BBnum BBid ref try hnd                 weight    [IL range]     [jump]      [EH region]         [flags]
--------------------------------------------------------------------------------------------------------------------------------------
BB23 [0016]  1                             1     [000..01B)-> BB24 (always)                     i 
BB24 [0018]  2                             1     [036..03C)        (return)                     i 
--------------------------------------------------------------------------------------------------------------------------------------

***************  Exception Handling table
index  eTry, eHnd
  0  ::            - Try at BB02..BB07 [016..038), Finally at BB08..BB08 [038..04D)

New BlockSet epoch 1, # of blocks (including unused BB00): 1, bitset array size: 1 (short)


----------- Statements (and blocks) added due to the inlining of call [000080] -----------

Inlinee method body:New Basic Block BB25 [0020] created.
EH#0: New last block of handler: BB25

Convert bbJumpKind of BB24 to BBJ_NONE
fgInlineAppendStatements: no gc ref inline locals.

--------------------------------------------------------------------------------------------------------------------------------------
BBnum BBid ref try hnd                 weight    [IL range]     [jump]      [EH region]         [flags]
--------------------------------------------------------------------------------------------------------------------------------------
BB23 [0016]  1     0                       1     [040..041)-> BB24 (always)    H0               i 
BB24 [0018]  2     0                       1     [040..041)                    H0               i 
--------------------------------------------------------------------------------------------------------------------------------------

------------ BB23 [040..041) -> BB24 (always), preds={} succs={BB24}

------------ BB24 [040..041), preds={} succs={BB25}

***** BB24, stmt 1
               [000140] ------------              *  STMT      void  (IL 0x040...  ???)
               [000137] --CXG-------              |  /--*  IND       ref   
               [000135] ------------              |  |  |  /--*  CNS_INT   int    0 Fseq[s_specificPool]
               [000136] --CXG-------              |  |  \--*  ADD       byref 
               [000134] H-CXG-------              |  |     \--*  CALL help byref  HELPER.CORINFO_HELP_GETSHARED_GCSTATIC_BASE_DYNAMICCLASS
               [000130] #----------- arg0         |  |        +--*  IND       long  
               [000129] ------------              |  |        |  \--*  CNS_INT(h) long   0x2408c12e7b0 cid/mid
               [000131] ------------ arg1         |  |        \--*  CNS_INT   int    58
               [000139] -ACXG-------              \--*  ASG       ref   
               [000138] D------N----                 \--*  LCL_VAR   ref    V08 tmp2         

-------------------------------------------------------------------------------------------------------------------

Return expression for call at [000080] is
               [000141] ------------              *  LCL_VAR   ref    V08 tmp2         
Successfully inlined ArrayPool`1:get_Shared():ref (66 IL bytes) (depth 1) [aggressive inline attribute]
--------------------------------------------------------------------------------------------

INLINER: during 'fgInline' result 'success' reason 'aggressive inline attribute' for 'Stream:CopyTo(ref,int):this' calling 'ArrayPool`1:get_Shared():ref'
INLINER: during 'fgInline' result 'success' reason 'aggressive inline attribute'

Replacing the return expression placeholder [000082] with [000141]
               [000082] --C---------              *  RET_EXPR  ref   (inl return from call [000141])

Inserting the inline return expression
               [000141] ------------              *  LCL_VAR   ref    V08 tmp2         

**** Late devirt opportunity
               [000085] --C-G-------              *  CALLV ind void   ArrayPool`1.Return
               [000141] ------------ this in rcx  +--*  LCL_VAR   ref    V08 tmp2         
               [000083] ------------ arg1         +--*  LCL_VAR   ref    V03 loc0         
               [000084] ------------ arg2         \--*  CNS_INT   int    0

impDevirtualizeCall: no type available (op=LCL_VAR)
*************** After fgInline()
```

**Comment by AndyAyersMS** [comment]
Yeah it's the same sort of issue as mentioned over in #9908. If a method like `get_Shared` does some jit-evaluable case analysis to return one of a number of objects exact/final types we will fail to devirtualize because we'll have spilled the return value.

Fix should hopefully not be too bad... will let you know in a bit.

**Comment by benaadams** [comment]
Looks like get the same without the if, so only a single type returned
```csharp
private static TlsOverPerCoreLockedStacksArrayPool<T> s_specificPool = new TlsOverPerCoreLockedStacksArrayPool<T>();

public static ArrayPool<T> Shared
{
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    get
    {
        return s_specificPool;
    }
}
```

```
Invoking compiler for the inlinee method ArrayPool`1:get_Shared():ref :
IL to import:
IL_0000  7e 5c 04 00 0a    ldsfld       0xA00045C
IL_0005  2a                ret         

INLINER impTokenLookupContextHandle for ArrayPool`1:get_Shared():ref is 0x0000029483F3BEA9.
...
Replacing the return expression placeholder [000082] with [000111]
               [000082] --C---------              *  RET_EXPR  ref   (inl return from call [000111])

Inserting the inline return expression
               [000111] --CXG-------              *  IND       ref   
               [000109] ------------              |  /--*  CNS_INT   int    0 Fseq[s_specificPool]
               [000110] --CXG-------              \--*  ADD       byref 
               [000108] H-CXG-------                 \--*  CALL help byref  HELPER.CORINFO_HELP_GETSHARED_GCSTATIC_BASE_DYNAMICCLASS
               [000104] #----------- arg0               +--*  IND       long  
               [000103] ------------                    |  \--*  CNS_INT(h) long   0x29482c7e7b0 cid/mid
               [000105] ------------ arg1               \--*  CNS_INT   int    58

**** Late devirt opportunity
               [000085] --C-G-------              *  CALLV ind void   ArrayPool`1.Return
               [000111] --CXG------- this in rcx  +--*  IND       ref   
               [000109] ------------              |  |  /--*  CNS_INT   int    0 Fseq[s_specificPool]
               [000110] --CXG-------              |  \--*  ADD       byref 
               [000108] H-CXG-------              |     \--*  CALL help byref  HELPER.CORINFO_HELP_GETSHARED_GCSTATIC_BASE_DYNAMICCLASS
               [000104] #----------- arg0         |        +--*  IND       long  
               [000103] ------------              |        |  \--*  CNS_INT(h) long   0x29482c7e7b0 cid/mid
               [000105] ------------ arg1         |        \--*  CNS_INT   int    58
               [000083] ------------ arg1         +--*  LCL_VAR   ref    V03 loc0         
               [000084] ------------ arg2         \--*  CNS_INT   int    0


impDevirtualizeCall: no type available (op=IND)
```
Don't think I see a temp this time?

**Comment by AndyAyersMS** [comment]
Now you are running into limitations with the jit deducing the types of static fields (same issue we hit in #9276). This should work ok if jitted.

We could generalize the case analysis done in `gtGetClassHandle` to match more static expansion patterns, but I've been reluctant to do that since I hoped to capture the the relevant info in other ways.

Also, my "quick fix" can get a type for the return, but it's just the declared return type. The jit will have to realize that only one return is actually reachable to safely sharpen the type. So it is not quite as simple as I'd hoped, though perhaps still doable.

**Comment by AndyAyersMS** [comment]
Just finishing up some jit changes which can handle cases like this. Should have a PR out soon.

Because this sort of devirtualization happens "late" the devirtualization doesn't yet lead to inlining.

**Comment by benaadams** [comment]
`Stream:CopyTo(ref,int):this` devirtualizes `Rent` and `Return`

`Stream:Read(struct):int:this` doesn't devirtualize the `Rent`, but does devirtualize the `Return`

`**** Late devirt opportunity` only pops up for the  `Return`

Changing the non-generic pools from properties to fields causes less to devirtualize (one less inline event to trigger on?)

**Comment by AndyAyersMS** [comment]
A "late devirt" is a devirtualization enabled by inlining sharpening a return value's type. So it will kick in for  examples like `ArrayPool.Shared.Rent`.

It would be odd if the version of `Shared` you are using is inlined and `Rent` doesn't subsequently devirtualize. If you save off the jit dump somehwere I'll see if we left enough clues behind to figure out what happened.

The jit's propagation of types from static field accesses is still hit and miss, unfortunately, so you will often get better results if the static field access is wrapped by a method or property.

I really should fix this but am not sure when I'll have the time.

**Comment by benaadams** [comment]
https://aoa.blob.core.windows.net/aspnet/jitDump.txt

**Comment by AndyAyersMS** [comment]
The sharpening happens but is blocked by a copy.
```
lvaGrabTemp returning 9 (V09 tmp3) (a long lifetime temp) called for Inline return value spill temp.
lvaSetClass: setting class for V09 to (0000017DD4DBCE68) ArrayPool`1 
lvaUpdateClass: Updating class for V09 from (0000017DD4DBCE68) ArrayPool`1 to (0000017DD4DBCFD0) TlsOverPerCoreLockedStacksArrayPool`1 

Inserting the inline return expression
               [000129] ------------              *  LCL_VAR   ref    V09 tmp3 `

...
***** BB16, stmt 2
               [000010] ------------              *  STMT      void  (IL   ???...  ???)
               [000129] ------------              |  /--*  LCL_VAR   ref    V09 tmp3         
               [000009] -AC---------              \--*  ASG       ref   
               [000008] D------N----                 \--*  LCL_VAR   ref    V06 tmp0         

***** BB16, stmt 3
               [000017] ------------              *  STMT      void  (IL   ???...  ???)
               [000013] --C-G-------              |  /--*  CALLV ind ref    ArrayPool`1.Rent
               [000011] ------------ this in rcx  |  |  +--*  LCL_VAR   ref    V06 tmp0 
```
Suspect if you rewrote
```
byte[] buffer = ArrayPool<byte>.Shared.Rent(destination.Length);
```
as
```
int len = destination.Length;
byte[] buffer = ArrayPool<byte>.Shared.Rent(len);
```
It would devirtualize.

**Comment by benaadams** [comment]
In attached dump seems to sharpen it; but then not trigger devirt (for Rent)
```
lvaUpdateClass: Updating class for V09 from (0000017DD4DBCE68) ArrayPool`1 to (0000017DD4DBCFD0) TlsOverPerCoreLockedStacksArrayPool`1 
```

**Comment by benaadams** [comment]
> It would devirtualize.

Seems to work 😄

**Comment by benaadams** [comment]
Every call to ArrayPool in crossgen'd System.Private.CoreLib is now devitualized to
```asm
call     TlsOverPerCoreLockedStacksArrayPool`1:Rent(int):ref:this
call     TlsOverPerCoreLockedStacksArrayPool`1:Return(ref,bool):this
```
Slight size regression
```asm
Total bytes of diff: 252 (0.01% of base)
    diff is a regression.

Total byte diff includes 178 bytes from reconciling methods
        Base had    0 unique methods,        0 unique bytes
        Diff had    3 unique methods,      178 unique bytes

Top file regressions by size (bytes):
         252 : System.Private.CoreLib.dasm (0.01% of base)

1 total files with size differences (0 improved, 1 regressed), 0 unchanged.

Top method regessions by size (bytes):
         110 : System.Private.CoreLib.dasm - ArrayPools:.cctor() (0/1 methods)
          34 : System.Private.CoreLib.dasm - ArrayPools:get_BytePool():ref (0/1 methods)
          34 : System.Private.CoreLib.dasm - ArrayPools:get_CharPool():ref (0/1 methods)
          16 : System.Private.CoreLib.dasm - BinaryWriter:Write(struct):this (3 methods)
          12 : System.Private.CoreLib.dasm - Stream:Write(struct):this

Top method improvements by size (bytes):
         -12 : System.Private.CoreLib.dasm - ArrayPool`1:.cctor() (3 methods)
          -3 : System.Private.CoreLib.dasm - Stream:Read(struct):int:this

27 total methods with size differences (2 improved, 25 regressed), 16743 unchanged.
```

Thank you @AndyAyersMS

**Comment by benaadams** [comment]
@dotnet-bot Tizen armel Cross Checked Innerloop Build and Test

**Comment by jkotas** [comment]
Does it improve throughput?

File: src/mscorlib/shared/System/IO/BinaryWriter.cs [hunk 1]
```diff
@@ -398,11 +398,12 @@ public virtual void Write(ReadOnlySpan<byte> buffer)
             }
             else
             {
-                byte[] array = ArrayPool<byte>.Shared.Rent(buffer.Length);
+                int length = buffer.Length;
+                byte[] array = ArrayPool<byte>.Shared.Rent(length);
```
Comment (jkotas) [inline src/mscorlib/shared/System/IO/BinaryWriter.cs]: This sensitivity to the exact callsite pattern is pretty unfortunate.

**Comment by jkotas** [comment]
I have been wondering whether we should have a static methods on the `ArrayPool` that just call the `Shared` pool; or are optimized even further. The virtual generic abstraction over malloc/free is unfortunate.

File: src/mscorlib/shared/System/IO/BinaryWriter.cs (hunk 1 above)
Comment (benaadams) [reply src/mscorlib/shared/System/IO/BinaryWriter.cs]: Issue for it: https://github.com/dotnet/coreclr/issues/15783

**Comment by benaadams** [comment]
> Does it improve throughput?

Some looks to be 7% - 10% of time
```csharp
[Benchmark(OperationsPerInvoke = OperationsPerInvoke)]
public static void RentReturn()
{
    for (var i = 0; i < OperationsPerInvoke; i++)
    {
        var buffer = ArrayPool<byte>.Shared.Rent(32);
        ArrayPool<byte>.Shared.Return(buffer);
    }
}

[Benchmark(OperationsPerInvoke = OperationsPerInvoke)]
public static void DoubleRentReturn()
{
    for (var i = 0; i < OperationsPerInvoke; i++)
    {
        var buffer1 = ArrayPool<byte>.Shared.Rent(32);
        var buffer2 = ArrayPool<byte>.Shared.Rent(32);
        ArrayPool<byte>.Shared.Return(buffer1);
        ArrayPool<byte>.Shared.Return(buffer2);
    }
}
```
Pre
```
           Method |      Mean |     Error |    StdDev |
----------------- |----------:|----------:|----------:|
       RentReturn |  35.68 ns | 0.0412 ns | 0.0365 ns |
 DoubleRentReturn | 122.95 ns | 0.4411 ns | 0.4126 ns |
```

Post
```
           Method |      Mean |     Error |    StdDev |
----------------- |----------:|----------:|----------:|
       RentReturn |  32.38 ns | 0.0341 ns | 0.0319 ns |
 DoubleRentReturn | 113.92 ns | 0.0631 ns | 0.0560 ns |
```

**Comment by benaadams** [comment]
@dotnet-bot test OSX10.12 x64 Checked Innerloop Build and Test 
@dotnet-bot test Tizen armel Cross Checked Innerloop Build and Test

**Comment by jkotas** [comment]
This will break people who are hacking around ArrayPool "leaks" using private reflection: https://github.com/dotnet/corefx/issues/25841#issue-281191035

**Comment by stephentoub** [comment]
> This will break people who are hacking around ArrayPool "leaks" using private reflection

I don't think we should be concerned about such "breaks"; limiting ourselves in such a way would prevent a huge number of improvements across the whole stack.

If we want to enable replacing the shared pool, we should do so explicitly with API/config/etc. and not promote such private reflection hacks.

**Comment by jkotas** [comment]
I agree that the private reflection hack is unsupported.

The problem is that the shared ArrayPool implementation leaks in number of cases, and the private reflection hack is the only way to workaround the leak. (BTW: This leak is one of the reasons why we have not ported anything improvements to .NET Framework yet.)

I think we need to have an answer for the leaking before we can take this change.

**Comment by benaadams** [comment]
Should this be marked as [blocked]? 😢 

As an aside the two pools can grow to (>= 64 procs):
`byte` pool 871 MB 
`char` pool at 1741 MB
For total of 2613 MB; should there be a lower level for 32 bit processes?

Although; perhaps more immediate may be limiting the number of `ThreadStatic`s that can be created for people that have gone wild and either set the threadpool min threads really high or spew lots of long running Tasks https://github.com/Microsoft/msbuild/pull/2384

**Comment by benaadams** [comment]
Added https://github.com/dotnet/coreclr/issues/15797 and https://github.com/dotnet/coreclr/issues/15796

File: src/mscorlib/shared/System/Buffers/ArrayPool.cs [hunk 2]
```diff
@@ -2,8 +2,25 @@
...
 {
+    internal static class ArrayPools
+    {
+        // Shared specialized ArrayPool<T> types created in non-generic class 
+        // to assist in devirtualization of ArrayPool<T>.Shared.
+        // The type in generic static field expansion is complex and 
+        // not picked up at AoT/crossgen time so won't currently devirtualize 
+        // when direct fields of ArrayPool<T> as part of crossgen.
+        // The general T pool should however devirtualize at Jit time.
+        public static TlsOverPerCoreLockedStacksArrayPool<byte> BytePool { [MethodImpl(MethodImplOptions.AggressiveInlining)] get; } =
```
Comment (jkotas) [inline src/mscorlib/shared/System/Buffers/ArrayPool.cs]: This should not be needed after https://github.com/dotnet/coreclr/pull/17078

Comment (benaadams) [reply src/mscorlib/shared/System/Buffers/ArrayPool.cs]: Only changes needed would be the `.Length` => `length` changes? (due to https://github.com/dotnet/coreclr/issues/15783)

**Comment by benaadams** [comment]
Changed to only use simple length in the Rent calls to work around  #15783

As the rest should be taken care of by  #17078

**Comment by benaadams** [comment]
(Jit - CoreFx) failures are `System.Threading.Tasks.ValueTask`1.AsTask()` assert failure in Pipelines tests

**Comment by benaadams** [comment]
Updated summary for reduced scope.

/cc @jkotas this should be a safe change now? Is is valid for 2.1?

File: src/mscorlib/shared/System/Collections/Generic/ValueListBuilder.cs [hunk 3]
```diff
@@ -60,7 +60,10 @@ public void Dispose()
 
         private void Grow()
         {
-            T[] array = ArrayPool<T>.Shared.Rent(_span.Length * 2);
+            // Use simple variable to call ArrayPool.Shared.Rent to allow devirtualization
+            // https://github.com/dotnet/coreclr/issues/15783
+            var length = _span.Length * 2;
```
Comment (stephentoub) [inline src/mscorlib/shared/System/Collections/Generic/ValueListBuilder.cs]: var => int

**Comment by stephentoub** [comment]
It's unfortunate seeing all of these call sites made more complicated as a workaround for something that should be doable in the JIT.  Are you able to measure a throughput improvement for any of these call sites (not a microbenchmark for Rent or Return)?

**Comment by benaadams** [comment]
> Are you able to measure a throughput improvement for any of these call sites (not a microbenchmark for Rent or Return)?

It saves about 1.65 ns per call; which is half an `await` on a completed task, or 1/10th of a dictionary `ContainsKey` on an `int` key

Don't think the paths are heavily trodden; default methods that should be overriden and less used areas like `RegistryKey` and `BinaryWriter`. 

`ValueListBuilder` and `ValueStringBuilder` might be more valuable.

Just checking now to see how things devirtualize currently

**Comment by benaadams** [comment]
Hmm... doesn't seem to be devirtualizing at all currently

File: src/mscorlib/shared/System/Buffers/ArrayPool.cs (hunk 2 above)
Comment (benaadams) [reply src/mscorlib/shared/System/Buffers/ArrayPool.cs]: Its stored in a `ArrayPool` type so the Jit can't workout its type; it needs to be stored in a `TlsOverPerCoreLockedStacksArrayPool` backing field so the Jit can see the type

File: src/mscorlib/shared/System/Buffers/ArrayPool.cs [hunk 4]
```diff
@@ -2,6 +2,9 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+using System.Runtime.CompilerServices;
```
Comment (jkotas) [inline src/mscorlib/shared/System/Buffers/ArrayPool.cs]: These should not be needed.

**Comment by benaadams** [comment]
Devirtualization gets blocked by the generic in the generic 😢 

```csharp
       call     CORINFO_HELP_GETSHARED_GCSTATIC_BASE_DYNAMICCLASS
       mov      rdx, gword ptr [rax]
       lea      rcx, bword ptr [rdi+8]
       call     CORINFO_HELP_ASSIGN_REF
       mov      rcx, rbx
       mov      edx, 15
       call     CORINFO_HELP_GETSHARED_GCSTATIC_BASE_DYNAMICCLASS
       mov      rbx, rax
       mov      rcx, gword ptr [rbx]
       mov      edx, 256
       mov      rax, qword ptr [rcx]
       mov      rax, qword ptr [rax+64]
       call     gword ptr [rax+32]ArrayPool`1:Rent(int):ref:this
```

**Comment by benaadams** [comment]
Wrong button

**Comment by benaadams** [comment]
This needs the original work around with some specialized types e.g. `<byte>` and `<char>` going in a non-generic class to work :-/

**Comment by jkotas** [comment]
Or fix in this JIT so that it just works...

**Comment by benaadams** [comment]
> Or fix in the JIT so that it just works...

That would also improve a lot more areas that this :) Like `Array.Empty<T>()` in a generic

**Comment by benaadams** [comment]
Want to mark as no-merge?

**Comment by benaadams** [comment]
For completeness have updated to what you need to do currently to get
```asm
       call     StreamHelpers:ValidateCopyToArgs(ref,ref,int)
       mov      rcx, qword ptr [(reloc)]
       mov      edx, 439
       call     CORINFO_HELP_GETSHARED_GCSTATIC_BASE
       mov      r14, rax
       mov      rcx, gword ptr [r14+936]
       mov      edx, ebx
       cmp      dword ptr [rcx], ecx
       call     TlsOverPerCoreLockedStacksArrayPool`1:Rent(int):ref:this   ; Devirtualized
       mov      gword ptr [rbp-28H], rax
G_M13776_IG03:
       jmp      SHORT G_M13776_IG05
G_M13776_IG04:
       mov      rcx, rdi
       mov      rdx, gword ptr [rbp-28H]
       mov      r9d, eax
       xor      r8d, r8d
       mov      rax, qword ptr [rdi]
       mov      rax, qword ptr [rax+96]
       call     qword ptr [rax+40]Stream:Write(ref,int,int):this
G_M13776_IG05:
       mov      r9, gword ptr [rbp-28H]
       mov      r9d, dword ptr [r9+8]
       mov      rcx, rsi
       mov      rdx, gword ptr [rbp-28H]
       xor      r8d, r8d
       mov      rax, qword ptr [rsi]
       mov      rax, qword ptr [rax+96]
       call     qword ptr [rax+16]Stream:Read(ref,int,int):int:this
       test     eax, eax
       jne      SHORT G_M13776_IG04
G_M13776_IG06:
       mov      rcx, gword ptr [r14+936]
G_M13776_IG07:
       mov      rdx, gword ptr [rbp-28H]
       xor      r8d, r8d
       cmp      dword ptr [rcx], ecx
       call     TlsOverPerCoreLockedStacksArrayPool`1:Return(ref,bool):this ; Devirtualized
       nop 
```
```
Total bytes of diff: 180 (0.01% of base)
    diff is a regression.

Total byte diff includes 178 bytes from reconciling methods
        Base had    0 unique methods,        0 unique bytes
        Diff had    3 unique methods,      178 unique bytes

Top file regressions by size (bytes):
         180 : System.Private.CoreLib.dasm (0.01% of base)

1 total files with size differences (0 improved, 1 regressed), 0 unchanged.

Top method regessions by size (bytes):
         110 : System.Private.CoreLib.dasm - ArrayPool:.cctor() (0/1 methods)
          34 : System.Private.CoreLib.dasm - ArrayPool:get_ByteArrayPool():ref (0/1 methods)
          34 : System.Private.CoreLib.dasm - ArrayPool:get_CharArrayPool():ref (0/1 methods)
          16 : System.Private.CoreLib.dasm - BinaryWriter:Write(struct):this (3 methods)
          13 : System.Private.CoreLib.dasm - ValueListBuilder`1:Grow():this (2 methods)

Top method improvements by size (bytes):
         -62 : System.Private.CoreLib.dasm - ArrayPool`1:.cctor() (4 methods)
         -21 : System.Private.CoreLib.dasm - Stream:ReadAsync(struct,struct):struct:this
         -13 : System.Private.CoreLib.dasm - Stream:Read(struct):int:this
         -13 : System.Private.CoreLib.dasm - TextReader:Read(struct):int:this
         -13 : System.Private.CoreLib.dasm - TextReader:ReadBlock(struct):int:this

36 total methods with size differences (5 improved, 31 regressed), 17261 unchanged.
```

File: src/mscorlib/shared/System/Buffers/ArrayPool.cs (hunk 4 above)
Comment (benaadams) [reply src/mscorlib/shared/System/Buffers/ArrayPool.cs]: Is needed again

**Pull request closed without merging**